### DISCUSSION TITLE
Why does Bokeh 2.2.3 specify Jinja2 `>=2.7` instead of `^2.7`?

### DISCUSSION THREAD

**Discussion by hintron** [bokeh/bokeh#12626] [Internals and design (Q&A)] [answered]
Why does Bokeh 2.2.3 specify Jinja2 `>=2.7` instead of `^2.7` in pyproject.toml?

The problem I see is that when I import Bokeh via poetry, the `>= 2.7` requirement is also satisfied by Jinja2 `>= 3.0`, and 3.x breaks Bokeh 2.2.3. If, however, this requirement was changed to `^2.7`, I don't think this wouldn't happen.

See also https://python-poetry.org/docs/dependency-specification/#caret-requirements

Here's an example of Jinja2 itself not putting an upper bound on one of its dependencies, MarkupSafe: https://github.com/pallets/markupsafe/issues/286

**Comment by bryevdv** [comment, marked as answer]
> and 3.x breaks Bokeh

Can you elaborate? I don't believe this claim is actually true. I use Jinja 3.1.2 daily with no issues, and CI unit tests  [use Jinja 3.x on every push](https://github.com/bokeh/bokeh/actions/runs/3485480840/jobs/5831166248#step:3:786) as well.

**Reply by bryevdv** [reply]
It looks like some changes to support Jinja2 3.x were made here: https://github.com/bokeh/bokeh/issues/11175 which landed in Bokeh version 2.3.2. Are you using an older version of Bokeh? If so, I don't have a solution except to suggest upgrading, it is extremely unlikely we would or could issue a new point release for `2.3.x` at this point.

**Reply by hintron** [reply]
Yeah, I'm using Bokeh 2.2.3, so I guess I just barely missed out on that fix :) Sorry, I should have put down the version of Bokeh I was using. Thanks for the help; I will aim to upgrade to 2.3.x

**Reply by bryevdv** [reply]
2.2.x point release is even less likely :) We have the resources to look back 1-2 release levels for point releases, beyond that it is increasingly difficult to keep build machinery functional as time passes.

**Reply by hintron** [reply]
I understand, no problem. I'm just glad it's not an issue in later versions.
